### PR TITLE
reserve and reserve_exact

### PR DESCRIPTION
**Pull request by mbrubeck**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -248,12 +249,52 @@ impl<T> ThinVec<T> {
...
+        let len = self.len();
+        let old_cap = self.capacity();
+
+        // Ensure the new capacity is at least double, to guarantee exponential growth.
+        let min_cap = len.checked_add(additional).expect("capacity overflow");
+        let double_cap = old_cap.saturating_mul(2);
+        let new_cap = max(min_cap, double_cap);
+
+        let additional = new_cap - len;
+        self.reserve_exact(additional);
```
Comment (mbrubeck) [inline src/lib.rs]: Hopefully the redundant checks get optimized out; otherwise we could just copy/paste more code here.

File: src/lib.rs [hunk 2]
```diff
@@ -303,31 +344,6 @@ impl<T> ThinVec<T> {
...
-        // TODO: capacity overflow logic?
-        let old_cap = self.capacity();
-        let old_len = self.len();
-
-        if old_cap <= old_len {
-            let elem_size = mem::size_of::<T>();
-
-            // "Infinite" capacity for ZSTs
-            let new_cap = if elem_size == 0 { !0 } else 
-                          if old_cap == 0 { 4 } else
```
Comment (Gankro) [inline src/lib.rs]: This optimization was lost in translation

File: src/lib.rs (hunk 1 above)
Comment (Gankro) [inline src/lib.rs]: Definitely a "post 0.1" task.

File: src/lib.rs (hunk 2 above)
Comment (mbrubeck) [inline src/lib.rs]: Fixed.  The "minimum 4" optimization is now applied in `reserve`.  The "infinite size for ZSTs" optimization is now applied in `header_with_capacity`.

**Comment by mbrubeck** [comment]
Factored out some common code, and added a `shrink_to_fit` implementation that uses it.

**Comment by mbrubeck** [comment]
This is affected by #8.  The alignment argument to `realloc` will need to be changed when this branch is merged with that one.

**Comment by mbrubeck** [comment]
Merged the padding fix from #8 into this PR because they would conflict otherwise.  See the commit message or #8 for details.

**Comment by Gankro** [comment]
I don't think removing the `[inbounds]` from all accesses into the Vec is worth potentially saving some space for SIMD types (which are going to eat enough space to make the padding irrelevant).

**Comment by mbrubeck** [comment]
The padding change is not about saving space; it's about correctness.  The commit log gives an example (not involving SIMD types) where the padding causes the data to have the wrong alignment.

**Comment by Gankro** [comment]
Ugh damnit I had proven to myself this was correct but I definitely took a wrong turn. What do you think about just banning high alignment types?

**Comment by Gankro** [comment]
Or alternatively allocating empty instances in those cases.

**Comment by mbrubeck** [comment]
Fixed to restore `[inbounds]` when applicable.

**Comment by mbrubeck** [comment]
(Specifically, this uses `inbounds_gep` for types with up to double-word-size alignment, which typically includes all primitive types.)

File: src/lib.rs [hunk 3]
```diff
@@ -278,16 +282,56 @@ impl<T> ThinVec<T> {
...
+    /// May reserve more space than requested, to avoid frequent reallocations.
+    ///
+    /// Panics if the new capacity overflows `usize`.
+    ///
+    /// Re-allocates only if `self.capacity() < self.len() + additional`.
     pub fn reserve(&mut self, additional: usize) {
-        // TODO
+        let len = self.len();
+        let old_cap = self.capacity();
+        let min_cap = len.checked_add(additional).expect("capacity overflow");
```
Comment (Gankro) [inline src/lib.rs]: This check is the reason why `reserve_one_more` was split off from `reserve`. I'd toss in a FIXME for it. Not a blocker.

Comment (Gankro) [inline src/lib.rs]: In fact the "right" fix is probably just to inline `len == cap` checks at each user, *shrug*. We'll get it in post!

Comment (mbrubeck) [inline src/lib.rs]: Added a `len == cap` check before calls from `push` and `insert`.

File: src/lib.rs [hunk 4]
```diff
@@ -53,37 +52,42 @@ fn oom() -> ! { std::process::abort() }
...
-    let header_align = mem::align_of::<Header>();
+fn padding<T>() -> usize {
+    let alloc_align = alloc_align::<T>();
+    let header_size = mem::size_of::<Header>();
+
+    if header_size % alloc_align == 0 {
+        0
+    } else {
+        alloc_align - header_size % alloc_align
+    }
```
Comment (mbrubeck) [inline src/lib.rs]: Fixed.

**Pull request merged**